synapse/http/federation/matrix_federation_agent.py

```diff
@@ -190,6 +240,45 @@ def _route_matrix_uri(self, parsed_uri):
             target_port=port,
         ))
 
+    @defer.inlineCallbacks
+    def get_well_known(self, server_name):
+        # FIXME: add a cache
+
+        uri = b"https://%s/.well-known/matrix/server" % (server_name, )
+        logger.debug("Fetching %s", uri.decode("ascii"))
+        try:
+            response = yield make_deferred_yieldable(
+                self._well_known_agent.request(b"GET", uri),
+            )
+        except Exception as e:
+            logger.info(
+                "Connection error fetching %s: %s",
+                uri.decode("ascii"), e,
+            )
+            defer.returnValue(None)
+
+        body = yield make_deferred_yieldable(readBody(response))
+
+        logger.info("Response from .well-known: %s", body)
+
+        if response.code != 200:
+            logger.info(
+                "Error response %i from %s",
+                response.code, uri.decode("ascii"),
+            )
+            defer.returnValue(None)
+
+        content_types = response.headers.getRawHeaders(u'content-type')
+        if content_types is None:
+            raise Exception("no content-type header on .well-known response")
+        content_type, _opts = cgi.parse_header(content_types[-1])
+        if content_type != 'application/json':
+            raise Exception("content-type not application/json on .well-known response")
+        parsed_body = json.loads(body)
+        if not isinstance(parsed_body, dict) or "m.server" not in parsed_body:
+            raise Exception("invalid .well-known response")
+        defer.returnValue(parsed_body["m.server"].encode("ascii"))
```

Comment:
The idea here is that if a server returns a 200 for a `.well-known/matrix/server` then we die if its not valid?

Reply:
yes. the MSC says:

> The response must have a Content-Type of application/json, and must be valid JSON which follows the structure documented below. Otherwise, the request is aborted.